### ISSUE TITLE
[Suggestion] Nullable return instead of default datatype values.

### ISSUE THREAD

**Issue by DividedSE** [Zeugma440/atldotnet#109]
## The problem

The file I sent you in my previous bug report has only 1 tag of genre & every other field is empty.
But if try to fetch any int(TrackTotal) or DataTime(Year) it will return 0 & 01.01.0001 00:00:00 instead of null. so users might think that data is actually stored there.

## Suggestion 
Returning int? instead of int & DateTime? instead of DateTime may be suitable. 
Didn't test another datatype. so for any other nullable data types.

**Comment by Zeugma440** [comment]
Switching to nullable types is better for semantics than relying on default values, I agree with you.

However, if I do that without second thoughts, it will cause a major compatibility break, as all client apps will suddenly have to deal with `null` values spawning everywhere.

The solution I have in mind is to :
- Make public fields of `Track` nullable where values may not be present
- Keep feeding them default values by default, depending on a new setting (e.g. `Settings.nullAbsentValues = false`) to maintain compatibility with old code
- If that setting is set to `true`, these fields will be nulled

**Comment by DividedSE** [comment]
Your solution sounds good to me.
```
var theTrack = new Track(fileName ) //false Default. 
var theTrack = new Track(fileName, Settings.nullAbsentValues: true); // if someone wants to get null values.
```

**Comment by DividedSE** [comment]
Also Saving of Null value int & DateTime .
as track.Year = null //giving error
or something like track.Year.Remove(); ????

**Comment by DividedSE** [comment]
Any Update on this... Simply can't delete or add null to any datetime or int values.

**Comment by Zeugma440** [comment]
Sorry, I thought that was only an API issue.

- `Integer` fields can be removed by setting them to `0`
- `DateTime` fields can be removed by setting them to `DateTime.MinValue` (didn't work until now : 03c4e856cfadf6df8de2fae85aa38009c6e239b8 )

As for the API change, it's quite a big change to do, and I don't have any bandwidth to do that immediately.
I'll get back to you when I can.

**Comment by DividedSE** [comment]
I tested it. Yes DateTime.MinValue now actully works. But its still doesnt remove the value completly... 
Ex.- My Code -> track.Date = value??DateTime.MinValue;
test1-> value of 8/5/2021 make TDRC 2021-08-05
test2-> value of null make TDRC 2021 // which is not null ????

**Comment by Zeugma440** [comment]
> TYER which I never stored still get holds data from internal atl mapping? or something else ?

That's the problem. `Year` was the legacy property, and `Date` was added later, on top of it, without proper synch. You just reminded me of doing a proper cleaning that was long overdue.

=> I've merged both fields (see latest commit), which should solve the issue.

NB : Beware of compatibility breaks due to the removal of `IMetaDataIO.Year`. If you're only using `Track`, nothing changes.

**Comment by Zeugma440** [comment]
@DividedSE could you confirm year and date now work as expected ?

**Comment by DividedSE** [comment]
Sorry about that. I dont have Internet at my PC for couple of days. so I was unable to test that. I will try to download it with mobile data let it know within a day or two.

**Comment by DividedSE** [comment]
Is the problem resolved. TYER dont hold any data anymore after clering TDRC.

**Comment by Zeugma440** [comment]
Thanks for the confirmation !

I'll keep that issue open for the API change (aka Nullable return instead of default datatype values 😉 )

**Comment by Zeugma440** [comment]
Nullable non-`string` fields are available in today's v4.01 !
